pkg/server/handler/auth.go

```diff
@@ -627,3 +627,484 @@ func (h *PasswordHandler) Handle(payload *router.Payload, response *router.Respo
+	PrincipalID      string                 `mapstructure:"principal_id"`
+	ProviderAuthData map[string]interface{} `mapstructure:"provider_auth_data"`
+	Profile          skydb.Data             `mapstructure:"profile"`
+}
+
+func (payload *signupProviderPayload) Decode(data map[string]interface{}) skyerr.Error {
+	if err := mapstructure.Decode(data, payload); err != nil {
+		return skyerr.NewError(skyerr.BadRequest, "fails to decode the request payload")
+	}
+	return payload.Validate()
+}
+
+func (payload *signupProviderPayload) Validate() skyerr.Error {
+	if payload.Provider == "" {
+		return skyerr.NewInvalidArgument("empty provider", []string{"provider"})
+	}
+
+	if payload.PrincipalID == "" {
+		return skyerr.NewInvalidArgument("empty principal id", []string{"principal_id"})
+	}
+
+	return nil
+}
+
+// SignupProviderHandler create new user with provider information
+//
+// LoginProviderHandler receives parameters:
+//
+// * provider (string, required)
+// * principal_id (string, required)
+// * provider_auth_data (json object, optional)
+// * profile (json object, optional)
+//
+// curl -X POST -H "Content-Type: application/json" \
+//   -d @- http://localhost:3000/ <<EOF
+// {
+// 		"action": "auth:signup_provider",
+// 		"provider": "facebook",
+// 		"principal_id": "104174434987489953648",
+// 		"provider_auth_data": {},
+// 		"profile": {"email": "chima@skygeario.com"}
+// }
+// EOF
+// Response
+// if no connected user
+// 		return user and token
+// else
+// 		return skyerr.InvalidArgument
+
+type SignupProviderHandler struct {
+	TokenStore       authtoken.Store    `inject:"TokenStore"`
+	ProviderRegistry *provider.Registry `inject:"ProviderRegistry"`
+	HookRegistry     *hook.Registry     `inject:"HookRegistry"`
+	AssetStore       asset.Store        `inject:"AssetStore"`
+	AuthRecordKeys   [][]string         `inject:"AuthRecordKeys"`
+	AccessKey        router.Processor   `preprocessor:"accesskey"`
+	DBConn           router.Processor   `preprocessor:"dbconn"`
+	InjectPublicDB   router.Processor   `preprocessor:"inject_public_db"`
+	PluginReady      router.Processor   `preprocessor:"plugin_ready"`
+	preprocessors    []router.Processor
+}
+
+func (h *SignupProviderHandler) Setup() {
+	h.preprocessors = []router.Processor{
+		h.AccessKey,
+		h.DBConn,
+		h.InjectPublicDB,
+		h.PluginReady,
+	}
+}
+
+func (h *SignupProviderHandler) GetPreprocessors() []router.Processor {
+	return h.preprocessors
+}
+
+func (h *SignupProviderHandler) Handle(payload *router.Payload, response *router.Response) {
+	log.Debugf("Signup provider")
+	p := &signupProviderPayload{}
+	skyErr := p.Decode(payload.Data)
+	if skyErr != nil {
+		response.Err = skyErr
+		return
+	}
+
+	if payload.HasMasterKey() == false {
+		response.Err = skyerr.NewError(skyerr.PermissionDenied, "no permission to signup provider")
+		return
+	}
+
+	store := h.TokenStore
+	info := skydb.AuthInfo{}
+	user := skydb.Record{}
+	principalID := p.Provider + ":" + p.PrincipalID
```

Comment:
`p.Provider + ":" + p.PrincipalID` logic should belong to `signupProviderPayload`?